### PR TITLE
Implement `ramalama push` command for pushing models to remote reposi…

### PR DESCRIPTION
**Pull request by ericcurtin**
…tories

Added `push_cli` function to handle the `ramalama push` command. Updated `funcDict` to include the `push` command.
Ensured proper validation of local model existence before pushing. Added error handling for push failures with descriptive messages.

### CODE REVIEW

**Comment by ericcurtin** [comment]
Completely untested, mostly boilerplate, if you could tell me how this is wrong @tarilabs I'd appreciate it, just had a spare 10 minutes to throw this together.

**Comment by tarilabs** [comment]
I'm on limited connectivity on national holiday atm, I think from cli a metadata file (either yaml or json) is mandatory [(example)](https://github.com/containers/omlmd/blob/c5c2797ca0aa517a444f5ab93a89c387f2245352/e2e/test_cli.sh#L15) but can be an emptye one!

Or if instead you modify ramalama to allow use of the py wheels, in the omlmd py sdk the metadata parameters can be [optional](https://github.com/containers/omlmd/blob/c5c2797ca0aa517a444f5ab93a89c387f2245352/omlmd/helpers.py#L39) [(full example)](https://containers.github.io/omlmd/#push).

As soon as I get back to my laptop I shall make an escape flag for omlmd CLI to deliberately ignore metadata, at least for these initial iterations...

File: docs/source/markdown/ramalama-push.1.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+% ramalama-push 1
+
+## NAME
+ramalama - Push AI Model
+
+## SYNOPSIS
+**ramalama push** [*options*] *model*
+
+## DESCRIPTION
+Push specified AI Model
```
Comment (rhatdan) [inline docs/source/markdown/ramalama-push.1.md:10]: If this is only supported for OCI, you should specify this in the man page.

File: ramalama [hunk 2]
```diff
@@ -333,6 +333,48 @@ def pull_cli(ramalama_store, args, port):
...
+def push_oci(model, ramalama_store):
+    # Remove the prefix and extract target details
+    target = re.sub(r'^oci://', '', model)
+    registry, reference = target.split('/', 1)
+    registry, reference = ("docker.io", target) if "." not in registry else (
+        registry, reference)
+    reference_dir = reference.replace(":", "/")
+
+    # Validate the model exists locally
+    local_model_path = os.path.join(
```
Comment (ericcurtin) [inline ramalama:345]: To do: we should be able to push any model to oci registries, including huggingface and ollama

File: ramalama [hunk 3]
```diff
@@ -333,6 +333,48 @@ def pull_cli(ramalama_store, args, port):
 funcDict["pull"] = pull_cli
 
 
+def push_oci(model, ramalama_store):
+    # Remove the prefix and extract target details
+    target = re.sub(r'^oci://', '', model)
+    registry, reference = target.split('/', 1)
+    registry, reference = ("docker.io", target) if "." not in registry else (
```
Comment (rhatdan) [inline ramalama:340]: We should not force docker.io.  Users need to fully specify on push.

Comment (tarilabs) [reply ramalama:340]: We aren't forcing indeed,
this is more to comply with the fact if I indicate `my-artifact:v1` as a reference,
it is intended to be `docker.io/library/my-artifact:v1`.

I believe most user flow however are in the complete form, such as `quay.io/mmortari/my-artifact:v1` anyway.

If the reference convention is not desired, we can always make it display an error message and fail, for instance ~"you must supply OCI registry in your target reference".

wdyt?

Comment (rhatdan) [reply ramalama:340]: I would prefer the latter.

**Comment by ericcurtin** [comment]
@tarilabs I'd be happy to let you take on this PR in general, you know this better than me. We need to support a simple:

`ramalama push MODEL`

and that should push any model we have locally. We can have additional flags for more complex use cases also.

Feel free to open a new PR and we can close this one.

**Comment by tarilabs** [comment]
@ericcurtin let me add the `--no-metadata` flag to omlmd, then happy to take a contribution at this 👍

File: ramalama (hunk 3 above)
Comment (tarilabs) [reply ramalama:340]: @rhatdan ack 👍

Comment (tarilabs) [reply ramalama:340]: I cross checked. This is for consistency of Pull semantic for OCI reference, where if not specified (eg.: `mmortari/model:v1`) is intended as coming from docker.io.

**Review by tarilabs — APPROVED** [review]
thank you so much @ericcurtin et al., 

I've rebased due to conflict with default main branch,
and added as discussed.

with:
```
./ramalama list                                          
NAME                                                                MODIFIED        SIZE  
oci://quay.io/mmortari/gguf-py-example/v1/example.gguf              1 second ago    4.0K
```

now both:
```
./ramalama push oci://quay.io/mmortari/gguf-py-example/v1/example.gguf quay.io/mmortari/gguf-py-example:ramalama
```

and:
```
./ramalama push oci://quay.io/mmortari/gguf-py-example/v1/example.gguf oci://quay.io/mmortari/gguf-py-example:ramalama
```

works, demo:
Screenshot 2024-08-22 at 10 22 46

File: ramalama (hunk 3 above)
Comment (rhatdan) [reply ramalama]: That is the way Docker works but not Podman.  Podman allows you to set a priority on container registries.  We should not hard code a preference for storing at any container registry.  Especially not a fully free one.

Comment (tarilabs) [reply ramalama]: TIL, and thank you for the explanation; will adapt to it ! 👍

Comment (tarilabs) [reply ramalama]: 👉  290a52b536c1ee693b760551f7a8388ef0bd4b33

**Comment by rhatdan** [comment]
LGTM

**Pull request merged**